### PR TITLE
Simplify structure

### PR DESCRIPTION
**Pull request by ktrns**
- Update to Seurat 4, python3, and updated all related R packages
- Overall, the structure is simplified
- Per run, we focus on one normalisation, not two
- RLE plots are moved to below merging/integrating samples
- We now cover the case that no differentially expressed genes are found
- Colours are exported as meta-data
- Added plots for highest expressed genes, and the number of cells per sample per cluster
- Moved code for processing differentially expressed genes into functions (R/functions_degs.R)
- Fixed issue #58 
- Fixed issue #55 
- Fixed issue #31

### CODE REVIEW

File: scrnaseq.Rmd [hunk 1]
```diff
@@ -128,7 +135,7 @@ param$latent_vars = c()
 param$integrate_samples = list(method="standard", reference_dataset=1, dimensions=30)
 
 # Which normalisation should be used for analysis: RNA or SCT?
-param$normalisation_default = "SCT"
+param$norm = "SCT"
```
Comment (andpet0101) [inline scrnaseq.Rmd]: Can we do RNA as default?

File: scrnaseq.Rmd [hunk 2]
```diff
@@ -167,6 +174,7 @@ param$path_to_git = "."
 git_files_to_source = c("R/functions_io.R",
               "R/functions_plotting.R",
               "R/functions_analysis.R",
+              "r/functions_degs.R",
```
Comment (andpet0101) [inline scrnaseq.Rmd]: r => R

File: scrnaseq.Rmd [hunk 3]
```diff
@@ -167,6 +174,7 @@ param$path_to_git = "."
...
               "R/functions_util.R")
```
Comment (andpet0101) [inline scrnaseq.Rmd:178]: In functions_util, check_installed_packages, can you change paste => paste0 or remove the spaces

also in check_python, can you change py_available => py_module_available

File: scrnaseq.Rmd [hunk 4]
```diff
@@ -688,45 +721,62 @@ if (param$cc_remove) {
...
+  # This is a new normalisation method that replaces previous Seurat functions 'NormalizeData', 'FindVariableFeatures', and 'ScaleData'. 
+  # vignette: https://satijalab.org/seurat/v3.0/sctransform_vignette.html
+  # paper: https://www.biorxiv.org/content/10.1101/576827v2
+  # Normalised data end up here: sc@assays$SCT@data
+  # Note: For a single dataset where no integration is needed, all features can already be scaled here. 
+  #   Otherwise, it is enough to scale only the variable features.
+  # Note: It is not guaranteed that all genes are successfully normalised with SCTransform. 
+  #   Consequently, some genes might be missing from the SCT assay. 
+  #   See: https://github.com/ChristophH/sctransform/issues/27
+sc = purrr::map(list_names(sc), function(n) {
```
Comment (andpet0101) [inline scrnaseq.Rmd]: Intendation

File: scrnaseq.Rmd [hunk 5]
```diff
@@ -688,45 +721,62 @@ if (param$cc_remove) {
...
+  # vignette: https://satijalab.org/seurat/v3.0/sctransform_vignette.html
+  # paper: https://www.biorxiv.org/content/10.1101/576827v2
+  # Normalised data end up here: sc@assays$SCT@data
+  # Note: For a single dataset where no integration is needed, all features can already be scaled here. 
+  #   Otherwise, it is enough to scale only the variable features.
+  # Note: It is not guaranteed that all genes are successfully normalised with SCTransform. 
+  #   Consequently, some genes might be missing from the SCT assay. 
+  #   See: https://github.com/ChristophH/sctransform/issues/27
+sc = purrr::map(list_names(sc), function(n) { 
+  SCTransform(sc[[n]],
```
Comment (andpet0101) [inline scrnaseq.Rmd]: Sometimes, there will be many warnings. Is there something like a function to truncate, i.e. show only the first 5?

File: scrnaseq.Rmd [hunk 6]
```diff
@@ -737,117 +787,32 @@ p = patchwork::wrap_plots(p_list, ncol=2) + patchwork::plot_annotation("Variable
...
+## Combining multiple samples
 ```{r multi_dataset_integration_single, eval=(param$integrate_samples[["method"]]=="single"), include=(param$integrate_samples[["method"]]=="single")}
 # Called when there is only a single sample and no integration needed
+# Default assay is set automatically
 if (param$integrate_samples[["method"]]=="single") {
   sc = sc[[1]]
-  
-  # Just set default assay
-  DefaultAssay(sc) = param$normalisation_default
+  message("Your dataset contains 1 sample only. No merging/intregating required.")
```
Comment (andpet0101) [inline scrnaseq.Rmd:796]: No intregating?

File: scrnaseq.Rmd [hunk 7]
```diff
@@ -882,22 +847,35 @@ cat("## Integration of multiple datasets\n\n")
...
   
-  # Rerun SCTransform
-  DefaultAssay(sc) = "SCT"
-  min_cells_overall = max(purrr::map_int(param$feature_filter, function(f) {as.integer(f[["min_cells"]])}))
-  sc = suppressWarnings(SCTransform(sc, vars.to.regress=param$vars_to_regress, min_cells=min_cells_overall, verbose=FALSE, return.only.var.genes=FALSE))
+  } else if (param$norm == "SCT") {
+    # Rerun SCTransform
+    min_cells_overall = max(purrr::map_int(param$feature_filter, function(f) as.integer(f[["min_cells"]])))
+    sc = SCTransform(sc, 
+                     vars.to.regress=param$vars_to_regress,
```
Comment (andpet0101) [inline scrnaseq.Rmd]: Only use vars_to_regress instead of param$vars_to_regress. Same for feature_filter.

File: scrnaseq.Rmd [hunk 8]
```diff
@@ -1378,231 +1435,267 @@ if ((known_markers_n > 0) & any(idx_avgplot)) {
...
+# Note: By default, the function uses slot="data". Mast requires log data, so this is the correct way to do it.
+#   https://www.bioconductor.org/packages/release/bioc/vignettes/MAST/inst/doc/MAST-interoperability.html
+degs = suppressMessages(Seurat::FindAllMarkers(sc, assay="RNA", test.use="MAST",
+                                               only.pos=FALSE, min.pct=0.25, logfc.threshold=0.25,
+                                               latent.vars=param$latent_vars, verbose=FALSE, silent=TRUE))
+
+# If no degs are found, initialize the degs table so that further downstream (export) chunks run
+if (ncol(degs) == 0){
+    degs$avg_log2FC = degs$p_val = degs$p_val_adj = degs$pct.1 = degs$pct.2 = degs$cluster = as.numeric()
+  degs$gene = as.character()
```
Comment (andpet0101) [inline scrnaseq.Rmd:1474]: Intendation

File: scrnaseq.Rmd [hunk 9]
```diff
@@ -1378,231 +1435,267 @@ if ((known_markers_n > 0) & any(idx_avgplot)) {
...
-p = patchwork::wrap_plots(p_list, ncol=2) + 
-  patchwork::plot_annotation(title="UMAP, cells coloured by normalised gene expression data")
-p
+#### Violin plots (raw)
+```{r degs_plot_violin_raw, eval=degs_found, fig.height=fig_height_degs}
+if (degs_found) {
+  # Violin plot of raw gene expression counts
+  p_list = Seurat::VlnPlot(sc, features=genes_example, assay="RNA", slot="counts", combine=FALSE, pt.size=0.2)
+  for (i in seq(p_list)) p_list[[i]] = p_list[[i]] + AddStyle(title=genes_example_titles[i], 
+                                                              legend_title="Cell identity",
```
Comment (andpet0101) [inline scrnaseq.Rmd]: "Cell identity" => "Cluster"? Or better get rid of legend

File: scrnaseq.Rmd [hunk 10]
```diff
@@ -1378,231 +1435,267 @@ if ((known_markers_n > 0) & any(idx_avgplot)) {
...
-p
 ```
 
-### Heatmap
-```{r DEG_heatmap, fig.height=20}
-# Heatmap of top differentially expressed genes
-top = sc_markers_filt %>% 
-  dplyr::group_by(cluster) %>% 
-  dplyr::top_n(n=10, wt=p_val_adj_score)
-p = Seurat::DoHeatmap(sc, features=top$gene, group.colors=param$col_clusters) +
```
Comment (andpet0101) [inline scrnaseq.Rmd:1599]: Would it be possible to limit the number of cells per cluster? For a better performance

File: scrnaseq.Rmd [hunk 11]
```diff
@@ -1729,15 +1841,18 @@ We export the assay data, clustering, visualisation, marker genes and enriched p
 # top expressed genes
 sc = cerebroApp::getMostExpressedGenes(sc, column_cluster="seurat_clusters", column_sample="orig.ident", assay=Seurat::DefaultAssay(sc))
```
Comment (andpet0101) [inline scrnaseq.Rmd:1844]: Error in cerebroApp::getMostExpressedGenes(sc, column_cluster = "seurat_clusters",  : 
  unused arguments (column_cluster = "seurat_clusters", column_sample = "orig.ident")

but probably due to updated version of cerebroApp

**Review by andpet0101 — COMMENTED** [review]
Okay, looks good! Minor changes please

File: scrnaseq.Rmd (hunk 5 above)
Comment (ktrns) [reply scrnaseq.Rmd]: At some point, we had suppressWarnings, and we removed it for this PR. Should I add it back? At least in the chunk `multi_dataset_integration_merge`, we don't set `warning=FALSE`.

File: scrnaseq.Rmd (hunk 7 above)
Comment (ktrns) [reply scrnaseq.Rmd]: I don't understand this one. Why?

File: scrnaseq.Rmd (hunk 10 above)
Comment (ktrns) [reply scrnaseq.Rmd:1599]: I know this takes a bit, but I think the clients are interested in this very figure, an overview figure of all cells and all degs. So I would like to leave this for now.

File: scrnaseq.Rmd (hunk 11 above)
Comment (ktrns) [reply scrnaseq.Rmd:1844]: Well, what should I do with this one? Ignore? Would you like to re-run with an updated Cerebro package version for a test?

File: scrnaseq.Rmd (hunk 5 above)
Comment (ktrns) [reply scrnaseq.Rmd]: In the above mentioned chunk, I added `suppressWarnings`.

File: scrnaseq.Rmd (hunk 7 above)
Comment (ktrns) [reply scrnaseq.Rmd]: Missed that this is inside a function. Solved.

File: scrnaseq.Rmd (hunk 10 above)
Comment (ktrns) [reply scrnaseq.Rmd:1599]: Added a note that we keep this in mind for later.

File: scrnaseq.Rmd (hunk 11 above)
Comment (ktrns) [reply scrnaseq.Rmd:1844]: It run through for me. So I will deal with this, once I observe the error too.

**Pull request merged**